src/attributes/attributes.c

```diff
@@ -0,0 +1,356 @@
+        { CKA_OTP_SERVICE_IDENTIFIER, "CKA_OTP_SERVICE_IDENTIFIER", },
+        { CKA_OTP_SERVICE_LOGO, "CKA_OTP_SERVICE_LOGO", },
+        { CKA_OTP_SERVICE_LOGO_TYPE, "CKA_OTP_SERVICE_LOGO_TYPE", },
+        { CKA_GOSTR3410_PARAMS, "CKA_GOSTR3410_PARAMS", },
+        { CKA_GOSTR3411_PARAMS, "CKA_GOSTR3411_PARAMS", },
+        { CKA_GOST28147_PARAMS, "CKA_GOST28147_PARAMS", },
+        { CKA_HW_FEATURE_TYPE, "CKA_HW_FEATURE_TYPE", },
+        { CKA_RESET_ON_INIT, "CKA_RESET_ON_INIT", },
+        { CKA_HAS_RESET, "CKA_HAS_RESET", },
+        { CKA_PIXEL_X, "CKA_PIXEL_X", },
+        { CKA_PIXEL_Y, "CKA_PIXEL_Y", },
+        { CKA_RESOLUTION, "CKA_RESOLUTION", },
+        { CKA_CHAR_ROWS, "CKA_CHAR_ROWS", },
+        { CKA_CHAR_COLUMNS, "CKA_CHAR_COLUMNS", },
+        { CKA_COLOR, "CKA_COLOR", },
+        { CKA_BITS_PER_PIXEL, "CKA_BITS_PER_PIXEL", },
+        { CKA_CHAR_SETS, "CKA_CHAR_SETS", },
+        { CKA_ENCODING_METHODS, "CKA_ENCODING_METHODS", },
+        { CKA_MIME_TYPES, "CKA_MIME_TYPES", },
+        { CKA_MECHANISM_TYPE, "CKA_MECHANISM_TYPE", },
+        { CKA_REQUIRED_CMS_ATTRIBUTES, "CKA_REQUIRED_CMS_ATTRIBUTES", },
+        { CKA_DEFAULT_CMS_ATTRIBUTES, "CKA_DEFAULT_CMS_ATTRIBUTES", },
+        { CKA_SUPPORTED_CMS_ATTRIBUTES, "CKA_SUPPORTED_CMS_ATTRIBUTES", },
+        { CKA_ALLOWED_MECHANISMS, "CKA_ALLOWED_MECHANISMS", },
+        { CKA_VENDOR_DEFINED, "CKA_VENDOR_DEFINED", },
+};
+
+static const size_t attributes_types_len =
+        (sizeof(attributes_types)/sizeof(attributes_types[0]));
+
+/**
+ * Get single object attribute.
+ *
+ * @returns CK_RV Value returned by the PKCS#11 library. This will indicate
+ *   success or failure.
+ */
+CK_RV attributes_get(
+        /** [in] Valid PKCS11 session. */
+        CK_SESSION_HANDLE session,
+        /** [in] The object handle. */
+        CK_OBJECT_HANDLE object,
+        /** [in] The attribute type. */
+        CK_ATTRIBUTE_TYPE type,
+        /** [out] The output buffer. Set to NULL to get the required buffer
+         *    size in buf_len. */
+        uint8_t *buf,
+        /** [in, out] The size of buf. */
+        size_t *buf_len ) {
+    CK_ATTRIBUTE attr[] = { { type, NULL_PTR, (CK_ULONG)0 } };
+    CK_RV rv = CKR_OK;
+
+    if (CK_INVALID_HANDLE == session)
+        return CKR_ARGUMENTS_BAD;
+
+    if (CK_INVALID_HANDLE == object)
+        return CKR_ARGUMENTS_BAD;
+
+    if (NULL == buf_len)
+        return CKR_ARGUMENTS_BAD;
+
+    rv = funcs->C_GetAttributeValue(
+            session,
+            object,
+            (CK_ATTRIBUTE_PTR)&attr[0].type,
+            (CK_ULONG)1 );
+    if (rv != CKR_OK) {
+        goto attributes_get_1;
+    }
+
+    if (buf) {
+        if (*buf_len < (size_t)attr[0].ulValueLen) {
+            rv = CKR_ARGUMENTS_BAD;
+            goto attributes_get_1;
+        }
+    } else {
+        *buf_len = (size_t)attr[0].ulValueLen;
+        goto attributes_get_1;
+    }
+
+    attr[0].pValue = (CK_BYTE_PTR)buf;
```

Comment:
Now I see why you were allocating before. I think removing this second C_GetAttributeValue is correct because it will never be needed. If buf is null, we skip the second C_GetAttributeValue call. If buf is not null, then its value would already be filled by the first C_GetAttributeValue call (or that call will have returned some kind of error).

Reply:
Yes, but the allocation was an unnecessary because buf itself can be used directly for the purpose of C_GetAttributeValue(), so I think we should be good now.

Reply:
It's not redundant:

https://github.com/aws-samples/aws-cloudhsm-pkcs11-examples/blob/f30d3b362cc362c597e3b333125cd484755c1515/src/attributes/attributes.c#L164

The function can be called with (buf == NULL) to obtain the required attribute value size.  In the the first call, the buffer is NULL:

https://github.com/aws-samples/aws-cloudhsm-pkcs11-examples/blob/f30d3b362cc362c597e3b333125cd484755c1515/src/attributes/attributes.c#L181

Then before the next call, the buffer is set:

https://github.com/aws-samples/aws-cloudhsm-pkcs11-examples/blob/f30d3b362cc362c597e3b333125cd484755c1515/src/attributes/attributes.c#L200

I hope that explains the logic.

Reply:
Refactored and added comments per our discussion earlier today.